packages/jsii-kernel/lib/serialization.ts

```diff
@@ -0,0 +1,622 @@
+  [SerializationClass.Enum]: {
+    serialize(value, type, host): WireEnum | undefined {
+      host.debug('hullo');
+      if (nullAndOk(value, type)) { return undefined; }
+
+      if (typeof value !== 'string' && typeof value !== 'number') {
+        throw new Error(`Expected enum value, got ${JSON.stringify(value)}`);
+      }
+
+      host.debug('Serializing enum');
+
+      const enumType = type as spec.EnumType;
+      return { [TOKEN_ENUM]: `${enumType.fqn}/${host.findSymbol(enumType.fqn)[value]}` };
+    },
+    deserialize(value, type, host) {
+      if (nullAndOk(value, type)) { return undefined; }
+
+      if (!isWireEnum(value)) {
+        throw new Error(`Expected enum value, got ${JSON.stringify(value)}`);
+      }
+
+      return deserializeEnum(value, host.findSymbol);
+    },
+  },
+
+  // ----------------------------------------------------------------------
+  [SerializationClass.Array]: {
+    serialize(value, type, host) {
+      if (nullAndOk(value, type)) { return undefined; }
+
+      if (!Array.isArray(value)) {
+        throw new Error(`Expected array type, got ${JSON.stringify(value)}`);
+      }
+
+      const arrayType = type as spec.CollectionTypeReference;
+
+      return value.map(x => host.recurse(x, arrayType.collection.elementtype));
+    },
+    deserialize(value, type, host) {
+      if (nullAndOk(value, type)) { return undefined; }
+
+      if (!Array.isArray(value)) {
+        throw new Error(`Expected array type, got ${JSON.stringify(value)}`);
+      }
+
+      const arrayType = type as spec.CollectionTypeReference;
+
+      return value.map(x => host.recurse(x, arrayType.collection.elementtype));
+    },
+  },
+
+  // ----------------------------------------------------------------------
+  [SerializationClass.Map]: {
+    serialize(value, type, host) {
+      if (nullAndOk(value, type)) { return undefined; }
+
+      const mapType = type as spec.CollectionTypeReference;
+      return mapValues(value, v => host.recurse(v, mapType.collection.elementtype));
+    },
+    deserialize(value, type, host) {
+      if (nullAndOk(value, type)) { return undefined; }
+
+      const mapType = type as spec.CollectionTypeReference;
+      return mapValues(value, v => host.recurse(v, mapType.collection.elementtype));
+    },
+  },
+
+  // ----------------------------------------------------------------------
+  [SerializationClass.Struct]: {
+    serialize(value, type, host) {
+      if (nullAndOk(value, type)) { return undefined; }
+
+      if (typeof value !== 'object' || value == null) {
+        throw new Error(`Expected object, got ${JSON.stringify(value)}`);
+      }
+
+      // This looks odd, but if an object was originally passed in as a by-ref
+      // class, and it happens to conform to a datatype interface we say we're
+      // returning, return the actual object instead of the serialized value.
+      // NOTE: Not entirely sure yet whether this is a bug masquerading as a
```

Comment:
I think it's okay for the kernel to allow passing in structs by reference. It should Just Work I guess

Reply:
Yeah I think it can be backwards-compatible.